src/search/task_proxy.cc

```diff
@@ -1,11 +1,60 @@
 #include "task_proxy.h"
 
+#include "axioms.h"
+#include "state_registry.h"
+
 #include "task_utils/causal_graph.h"
 
 #include <iostream>
 
 using namespace std;
 
+State::State(const AbstractTask &task, const StateRegistry &registry,
+             const StateID &id, const PackedStateBin *buffer)
+    : task(&task), registry(&registry), id(id), buffer(buffer), values(nullptr),
+      state_packer(&registry.get_state_packer()),
+      num_variables(registry.get_num_variables()) {
+    assert(id != StateID::no_state);
+    assert(buffer);
+    assert(num_variables == task.get_num_variables());
+}
+
+State::State(const AbstractTask &task, const StateRegistry &registry,
+             const StateID &id, const PackedStateBin *buffer,
+             vector<int> &&values)
+    : State(task, registry, id, buffer) {
+    assert(num_variables == values.size());
+    this->values = make_shared<vector<int>>(move(values));
```

Comment:
Another point for potential performance optimizations: this is not really a dynamic vector, so the only real advantage over an int * is that it knows its size. But then num_variables is the size. This would certainly not be easy to change because we use vector elsewhere to pass the data into the state and also to pass the data out, so changing this would require many changes everywhere. But it's part of the more general discussion if our use of `vector` everywhere for non-dynamic size stuff is good.